mattermost-plugin/product/product.go

```diff
@@ -0,0 +1,323 @@
+	var err error
+	db, err = sqlstore.New(storeParams)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing the DB: %w", err)
+	}
+	if cfg.AuthMode == server.MattermostAuthMod {
+		layeredStore, err2 := mattermostauthlayer.New(cfg.DBType, sqlDB, db, logger, newMattermostAuthLayerPluginAPIImpl(services))
+		if err2 != nil {
+			return nil, fmt.Errorf("error initializing the DB: %w", err2)
+		}
+		db = layeredStore
+	}
+
+	psAPI := newPermissionsPluginAPIImpl(services)
+	permissionsService := mmpermissions.New(db, psAPI)
+	wsAPI := newWSPluginAPIImpl(services)
+
+	wsPluginAdapter := ws.NewPluginAdapter(wsAPI, auth.New(cfg, db, permissionsService), db, logger)
+
+	// TODO: create actual request.Context()
+	notifypluginAdapter := newPluginDeliveryAPIImpl(request.EmptyContext(), services)
+
+	backendParams := notifyBackendParams{
+		cfg:         cfg,
+		pluginAPI:   notifypluginAdapter,
+		store:       db,
+		permissions: permissionsService,
+		wsAdapter:   wsPluginAdapter,
+		serverRoot:  baseURL + "/boards",
+		logger:      logger,
+	}
+
+	var notifyBackends []notify.Backend
+
+	// TODO: add utility function or import
+	mentionsBackend, err := createMentionsNotifyBackend(backendParams)
+	if err != nil {
+		return nil, fmt.Errorf("error creating mention notifications backend: %w", err)
+	}
+	notifyBackends = append(notifyBackends, mentionsBackend)
+
+	// TODO: add utility function or import
+	subscriptionsBackend, err2 := createSubscriptionsNotifyBackend(backendParams)
+	if err2 != nil {
+		return nil, fmt.Errorf("error creating subscription notifications backend: %w", err2)
+	}
+	notifyBackends = append(notifyBackends, subscriptionsBackend)
+	mentionsBackend.AddListener(subscriptionsBackend)
+
+	params := server.Params{
+		Cfg:                cfg,
+		SingleUserToken:    "",
+		DBStore:            db,
+		Logger:             logger,
+		ServerID:           serverID,
+		WSAdapter:          wsPluginAdapter,
+		NotifyBackends:     notifyBackends,
+		PermissionsService: permissionsService,
+	}
+
+	server, err := server.New(params)
+	if err != nil {
+		fmt.Println("ERROR INITIALIZING THE SERVER", err)
+		return nil, err
+	}
+
+	routerService := services[app.RouterKey].(product.RouterService)
+	routerService.RegisterRouter("focalboard", server.GetRootRouter())
+	hooksService := services[app.HooksKey].(product.HooksService)
+
+	b := &Boards{
+		server:       server,
+		ws:           wsPluginAdapter,
+		mmConfig:     mmServer.Config,
+		hooksService: hooksService,
+	}
+
+	return b, nil
+}
+
+func (b *Boards) Start() error {
+	fmt.Printf("\n\n -------- STARTING BOARDS -------- \n\n\n")
+	// We need to register hooks after start, because the plugin environment
+	// starts after the channels product starts. And the boards product itself
+	// implements product.Hooks interface, it can be replaced by any other struct
+	// as this is just for demonstration purposes.
+	if err := b.hooksService.RegisterHooks("focalboard", b); err != nil {
```

Comment:
cc @wiggin77, Here I register `Boards` itself as the hooks provider. Any registered callback will be called once they are triggered from the server.